homeassistant/components/vesync/humidifier.py

```diff
@@ -0,0 +1,201 @@
+    if ha_mode is None:
+        _LOGGER.warning("Unknown mode '%s'", vs_mode)
+    return ha_mode
+
+
+def _get_vs_mode(ha_mode: str) -> str | None:
+    vs_mode = HA_TO_VS_MODE_MAP.get(ha_mode)
+    if vs_mode is None:
+        _LOGGER.warning("Unknown mode '%s'", ha_mode)
+    return vs_mode
+
+
+class VeSyncHumidifierHA(VeSyncBaseEntity, HumidifierEntity):
+    """Representation of a VeSync humidifier."""
+
+    # The base VeSyncBaseEntity has _attr_has_entity_name and this is to prevent
+    # none being appended to entity id.
+    _attr_name = ""
+
+    _attr_max_humidity = MAX_HUMIDITY
+    _attr_min_humidity = MIN_HUMIDITY
+    _attr_supported_features: HumidifierEntityFeature = HumidifierEntityFeature(
+        HumidifierEntityFeature.MODES
+    )
+
+    def __init__(
+        self,
+        humidifier: VeSyncHumidifierDevice,
+        coordinator: VeSyncDataCoordinator,
+    ) -> None:
+        """Initialize the VeSync humidifier device."""
+        super().__init__(humidifier, coordinator)
+        self.smarthumidifier = humidifier
+
+    @property
+    def available_modes(self) -> list[str]:
+        """Return the available mist modes."""
+        return [
+            ha_mode
+            for ha_mode in (
+                _get_ha_mode(vs_mode) for vs_mode in self.smarthumidifier.mist_modes
+            )
+            if ha_mode
+        ]
+
+    @property
+    def target_humidity(self) -> int:
+        """Return the humidity we try to reach."""
+        return self.smarthumidifier.config["auto_target_humidity"]
+
+    @property
+    def mode(self) -> str | None:
+        """Get the current preset mode."""
+        return _get_ha_mode(self.smarthumidifier.details["mode"])
+
+    @property
+    def extra_state_attributes(self) -> Mapping[str, Any]:
+        """Return the state attributes of the humidifier."""
+
+        attr = {}
+        for k, v in self.smarthumidifier.details.items():
+            if k in VS_TO_HA_ATTRIBUTES:
+                attr[VS_TO_HA_ATTRIBUTES[k]] = v
+            elif k in self.state_attributes:
+                attr[f"vs_{k}"] = v
+            else:
+                attr[k] = v
+        return attr
+
+    def set_humidity(self, humidity: int) -> None:
+        """Set the target humidity of the device."""
+        if humidity not in range(int(self.min_humidity), int(self.max_humidity + 1)):
+            raise ValueError(
+                "{humidity} is not between {self.min_humidity} and {self.max_humidity} (inclusive)"
+            )
+        if self.smarthumidifier.set_humidity(humidity):
+            self.schedule_update_ha_state()
+        else:
+            raise ValueError("An error occurred while setting humidity.")
+
+    def set_mode(self, mode: str) -> None:
+        """Set the mode of the device."""
+        if mode not in self.available_modes:
+            raise ValueError(
+                "{mode} is not one of the valid available modes: {self.available_modes}"
+            )
```

Comment:
also guarded by HA itself

Reply:
The service call SERVICE_SET_MODE is processed by [async_set_mode](https://github.com/home-assistant/core/blob/4019045e7ba2466ec574ae9accf06027b9a784d3/homeassistant/components/humidifier/__init__.py#L250) which does not have any checks, so it seems we do need to keep the checks.